resources/ext/cli/config-migration

```diff
@@ -0,0 +1,243 @@
+#!/bin/bash
+
+migration_command="puppetdb config-migration"
+
+#############
+# FUNCTIONS #
+#############
+
+# Backs up a file, if it hasn't already been backed up
+#
+# $1 - file to backup
+#
+# Example:
+#
+#   backupfile "/etc/myconfig"
+#
+function backupfile() {
+  # Create the global array if it doesn't already exist
+  if [ -z "$backupfile_list" ]; then
+    backupfile_list=()
+  fi
+
+  # We check the array to make sure the file isn't already backed up
+  if ! contains ${backupfile_list[@]} "$1"; then
+    local backup_path="$1.bak.$(date +%s)"
+    echo "Backing up $1 to ${backup_path} before making changes"
+    cp -p "$1" "$backup_path"
+
+    # Append to the array, so we don't need to back it up again later
+    backupfile_list+=($1)
+  fi
+}
+
+# This function searches for an element in an array returning 1 if it exists.
+#
+# $1 - array
+# $2 - item to search for
+#
+# Example:
+#
+#   myarray=('element1', 'element2')
+#   if contains ${myarray[@]}, "element1'; then
+#     echo "element1 exists in the array"
+#   fi
+#
+function contains() {
+  local n=$#
+  local value=${!n}
+  for ((i=1;i < $#;i++)); do
+    if [ "${!i}" == "${value}" ]; then
+      return 0
+    fi
+  done
+  return 1
+}
+
+# This function wraps sed for a line focused search and replace.
+#
+# * Makes sure its atomic by writing to a temp file and moving it _after_
+# * Escapes any forward slashes and ampersands on the RHS for you
+#
+# $1 - regexp to match
+# $2 - line to replace
+# $3 - file to operate on
+#
+# Example:
+#
+#    replaceline "^$mysetting.*" "mysetting = myvalue" /etc/myconfig
+#
+function replaceline {
+  backupfile "$3"
+  tmp="$3.tmp.$(date +%s)"
+  sed "s/$1/$(echo "$2" | sed -e 's/[\/&]/\\&/g')/g" "$3" > "$tmp"
+  mv "$tmp" "$3"
+}
+
+# This function copies the necessary PEM files from Puppet to the PuppetDB
+# SSL directory.
+#
+# This expects various environment variables to have already been set to work.
+#
+# Note: This does not respect directory structures, and will move files blindly
+function migrate_files() {
+  destination=$1
+  orig_file=$2
+  if [ ! -e "$orig_file" ]; then
+    echo "Warning: Unable to find all ini configuration files to copy"
+    echo
+    echo "  This tool requires the following files to exist:"
+    echo
+    echo "  * $orig_file"
+    echo
+    echo "  These files may be missing due to the fact that your host's Puppet"
+    echo "  certificates may not have been signed yet, probably due to the"
+    echo "  lack of a complete Puppet agent run. Try running puppet first, for"
+    echo "  example:"
+    echo
+    echo "      puppet agent --test"
+    echo
+    echo "  Afterwards re-run this tool then restart PuppetDB to complete the SSL"
+    echo "  setup:"
+    echo
+    echo "      ${migration_command} -f"
+    exit 1
+  fi
+  if [ -e "${destination}/${orig_file##*/}" ] ; then
+    if $force ; then
+      echo "Replacing file: ${destination}/${orig_file##*/} with ${orig_file}"
```

Comment:
Wonder if it might make sense to redirect status messages to stderr, i.e. `echo "some message" 1>&2`